rust/capture/src/utils.rs

```diff
@@ -36,3 +36,148 @@ pub fn uuid_v7() -> Uuid {
+                // Extract the four escape sequence bytes
+                let mut code_point_bytes: [u8; 4] =
+                    [bytes[i + 2], bytes[i + 3], bytes[i + 4], bytes[i + 5]];
+
+                // Convert the bytes to a string, then parse it into a u16 to check if it's a valid escape sequence
+                if let Ok(Ok(num)) =
+                    std::str::from_utf8(&code_point_bytes).map(|s| u16::from_str_radix(s, 16))
+                {
+                    if (0xD800..=0xDBFF).contains(&num) {
+                        // High surrogate without a following low surrogate
+                        if !is_next_low_surrogate(&bytes, i + 6) {
+                            // Replace with 'FFFD' (Unicode replacement character)
+                            code_point_bytes.copy_from_slice(b"FFFD");
+                        } else {
+                            // This is a high surrogate, and the next is a low one, so we should skip over both
+                            // without modification
+                            i += 12;
+                            continue;
+                        }
+                    } else if (0xDC00..=0xDFFF).contains(&num) {
+                        // Unpaired low surrogate - we know this, because if it had a preceding high surrogate,
+                        // we would have skipped over it in the previous iteration (above) - replace it
+                        code_point_bytes.copy_from_slice(b"FFFD");
+                    }
+                    // The unhandled else case is that this isn't part of a surrogate pair, so we don't need to do anything
+                } else {
+                    // if we couldn't parse those 4 bytes as a hex escape code, or couldn't go from that hex escape code to a u16, replace with 'FFFD'
+                    code_point_bytes.copy_from_slice(b"FFFD");
+                }
+                bytes[i + 2] = code_point_bytes[0];
+                bytes[i + 3] = code_point_bytes[1];
+                bytes[i + 4] = code_point_bytes[2];
+                bytes[i + 5] = code_point_bytes[3];
+                i += 6; // Move past the Unicode escape sequence
+                continue;
+            } else {
+                // Not enough bytes for a Unicode escape sequence, truncate the buffer to before the slash, then append the replacement characters
+                bytes.truncate(i);
+                bytes.extend_from_slice("\\uFFFD".as_bytes());
+                break; // We're done, we just replaced the last 4 bytes
+            }
+        }
+        i += 1;
+    }
+
+    // Convert bytes back to String
+    String::from_utf8(bytes)
+}
+
+fn is_next_low_surrogate(bytes: &[u8], start: usize) -> bool {
+    let len = bytes.len();
+    if start + 6 <= len && bytes[start] == b'\\' && bytes[start + 1] == b'u' {
+        let code_point_bytes = &bytes[start + 2..start + 6];
+        if let Ok(code_point_str) = std::str::from_utf8(code_point_bytes) {
```

Comment:
ah, that's cool... so `from_utf8` might throw but we don't care so we can only write code for the happy path. Nice, I approve :)

Reply:
much nicer than

```
try {
  blah()
} except(e: Unknown) {
   // we don't care
}
```

Reply:
No throwing in rust, it's all just enums, variants and pattern matching 🔫  (the closest rust has to "throw" is `panic!()`, which (handwaving) kills your process and prints a stack trace - you shouldn't try to catch panic, almost ever)

But yeah, very similar idea.